### PR TITLE
[CURA-1615] Plugins: Updating GCodeProfileReader and SliceInfo

### PR DESCRIPTION
**Pull request by thopiekar**
I got the plugin most of all working. At least the "successfully"
imported profile XY" dialog appears. But sadly the profile does not
appear in the list of profiles after that. I can only guess something is
blocking here.
Additionally it should be noted that G-Code exported from Cura 2.1.x
does not work here anymore on Cura 2.2.x.

### CODE REVIEW

**Comment by thopiekar** [comment]
I believed that there is something wrong with the name handling, but couldn't catch anything with:
```
diff --git a/UM/Settings/InstanceContainer.py b/UM/Settings/InstanceContainer.py
index 81023b2..aab6c78 100644
--- a/UM/Settings/InstanceContainer.py
+++ b/UM/Settings/InstanceContainer.py
@@ -228,6 +228,12 @@
             raise DefinitionNotFoundError("Could not find definition {0} required for instance {1}".format(definition_id, self._id))
         self._definition = definitions[0]
 
+        if "general" in parser:
+            if "name" in parser["general"]:
+                if not parser["general"]["name"] == self._name:
+                    Logger.log("w", "Found new name in parser: %s", parser["general"]["name"])
+                    self.setName(parser["general"]["name"])
+        
         if "metadata" in parser:
             self._metadata = dict(parser["metadata"])
```

**Comment by fieldOfView** [comment]
Without looking at your code (much), have a look at Simon's recent work on importing curaprofiles:
https://github.com/Ultimaker/Cura/blob/d74cecd9db9d45c7d9273bfa09ad16d48cc4696e/cura/CuraContainerRegistry.py#L141
* you may have to set the definition
* make sure imported names are unique

**Comment by thopiekar** [comment]
I guess the second is the problem here. I took yesterday a deeper look on
the generated Gcode files and saw that they come with their own
['general']['name'].
That's why I tried to to take the name from the parser and override it.

Sadly my codes doesn't take any effect.
So one solution could be to filter out the name when exporting the profile
into the G-code.

What do you think?
Am 17.06.2016 09:30 schrieb "Aldo Hoeben" <notifications@github.com>:

**Comment by Ghostkeeper** [comment]
Translating 2.1 profiles is a separate system. It happens before the profiles are loaded at all. The system simply reads the old setting files, translates them, saves new files and puts the old files somewhere else for reference. Then when the new files are in place, it continues with the start-up process.

**Comment by thopiekar** [comment]
So why doesn't it work to import old "2.1" gcode?

**Comment by Ghostkeeper** [comment]
Settings were renamed, changed from a boolean to an enum or had their effect changed. Apart from those, it could totally work. But I haven't taken a look at this pull request. I'm just answering a question fieldOfView posed.

**Comment by thopiekar** [comment]
@fieldOfView I looked at the code you mentioned and found something critical + see that there is "self._id = name", which I also added in my lines above I used to find the reason for the missing profile entry.

**Comment by thopiekar** [comment]
@Ghostkeeper: Ok, thanks :smile:

**Comment by thopiekar** [comment]
@fieldOfView I overlooked the return in the exception. Thought the following if might fail..

**Comment by thopiekar** [comment]
I also wonder about that when exporting a .curaprofile and .gcode does not include the same info. For example my .curaprofile
```
[general]
version = 2
name = Normal Quality
definition = fdmprinter

[metadata]
weight = -2
type = quality

[values]

```
Looks fine so far. I even didn't made any changes. But when looking at the gcode:
```
;SETTING_1 [general]\nversion = 2\nname = G-code-imported-profile\ndefinition = 
;SETTING_1 bq_hephestos_2\n\n[metadata]\n\n[values]\nwireframe_roof_drag_along =
;SETTING_1  0.8\nmeshfix_union_all = True\nraft_surface_speed = 30.0\ntravel_com
;SETTING_1 pensate_overlapping_walls_enabled = True\ncool_lift_head = False\nspe
;SETTING_1 ed_wall = 30.0\nlayer_height = 0.2\nwall_line_width_x = 0.4\naccelera
;SETTING_1 tion_support_roof = 3000\nretraction_hop = 0\nsupport_roof_height = 1
;SETTING_1 \nraft_surface_thickness = 0.1\nooze_shield_enabled = False\nskirt_mi
;SETTING_1 nimal_length = 30\nretraction_retract_speed = 25\nwireframe_printspee
;SETTING_1 d_down = 5\nmaterial_print_temperature = 210\nskin_outline_count = 0\
;SETTING_1 nsupport_angle = 50\nretraction_extra_prime_amount = 0\ndual = 0\nacc
;SETTING_1 eleration_support_infill = 3000\nacceleration_infill = 3000\nadhesion
;SETTING_1 _extruder_nr = 0\ntravel = 0\nspeed_print = 60\nmachine_use_extruder_
;SETTING_1 offset_to_offset_coords = True\njerk_wall_x = 20\nspeed_support = 60\
;SETTING_1 njerk_support = 20\ncool_fan_speed = 100.0\nprint_sequence = all_at_o
;SETTING_1 nce\nwireframe_height = 3\nxy_offset = 0\nmagic_mesh_surface_mode = n
;SETTING_1 ormal\ninfill_wipe_dist = 0.1\nexperimental = 0\njerk_print = 20\nraf
;SETTING_1 t_interface_line_spacing = 1.0\nraft_interface_line_width = 0.8\njerk
;SETTING_1 _skirt = 20\njerk_wall = 20\ntop_bottom_thickness = 1.2\nwireframe_up
;SETTING_1 _half_speed = 0.3\nspeed_support_infill = 60\nsupport_pattern = zigza
;SETTING_1 g\nraft_acceleration = 3000\nmachine_nozzle_size = 0.4\nprime_tower_w
;SETTING_1 ipe_enabled = False\nspeed = 0\nmeshfix = 0\nwireframe_printspeed = 5
;SETTING_1 \nbrim_width = 8.0\njerk_enabled = False\nmachine_nozzle_heat_up_spee
;SETTING_1 d = 2.0\nshell = 0\ndraft_shield_enabled = False\nwireframe_nozzle_cl
;SETTING_1 earance = 1\nskin_overlap_mm = 0.02\nmagic_fuzzy_skin_point_dist = 0.
;SETTING_1 8\nooze_shield_angle = 60\nwireframe_drag_along = 0.6\nsupport_connec
;SETTING_1 t_zigzags = True\nwireframe_enabled = False\nretraction_count_max = 9
;SETTING_1 0\nskin_overlap = 5\nsupport_roof_line_width = 0.4\nmeshfix_keep_open
;SETTING_1 _polygons = False\nmachine_nozzle_expansion_angle = 45\nspeed_prime_t
;SETTING_1 ower = 60\nretraction_min_travel = 0.8\nwireframe_top_jump = 0.6\npri
;SETTING_1 me_tower_position_x = 200\nswitch_extruder_retraction_speed = 20\njer
;SETTING_1 k_support_infill = 20\ndraft_shield_dist = 10\nmaterial_bed_temperatu
;SETTING_1 re = 0\nskirt_speed = 30\njerk_travel = 30\nspeed_wall_0 = 30.0\nspee
;SETTING_1 d_travel = 120\nsupport_conical_enabled = False\nwireframe_roof_fall_
;SETTING_1 down = 2\nmachine_disallowed_areas = []\nwall_0_inset = 0\nmachine_he
;SETTING_1 ated_bed = False\nmachine_center_is_zero = False\nmachine_nozzle_tip_
;SETTING_1 outer_diameter = 1\nraft_surface_jerk = 20\nmaterial_flow_temp_graph 
;SETTING_1 = [[3.5,200],[7.0,240]]\ninfill_line_width = 0.4\ntravel_avoid_other_
;SETTING_1 parts = True\ninfill_sparse_density = 20\nconical_overhang_angle = 50
;SETTING_1 \nmultiple_mesh_overlap = 0.15\nskin_alternate_rotation = False\ntop_
;SETTING_1 bottom_pattern = lines\nwall_thickness = 1.2\ncoasting_speed = 90\nma
;SETTING_1 chine_end_gcode = ; -- END GCODE --\n	M801        ; Custom GCODE to f
;SETTING_1 ire end print procedure\n	; -- end of END GCODE --\nspeed_wall_x = 60
;SETTING_1 .0\nraft_base_line_spacing = 3.0\nwireframe_roof_outer_delay = 0.2\nj
;SETTING_1 erk_wall_0 = 20\nskin_no_small_gaps_heuristic = True\nsupport_line_di
;SETTING_1 stance = 2.6666666666666665\nacceleration_layer_0 = 3000\nmaterial_fl
;SETTING_1 ow_dependent_temperature = False\nmachine_heat_zone_length = 16\nskir
;SETTING_1 t_line_count = 4\ninfill_pattern = grid\nmaterial = 0\nskirt_line_wid
;SETTING_1 th = 0.4\nalternate_extra_perimeter = False\nsupport_roof_pattern = c
;SETTING_1 oncentric\nmachine_height = 220\nwall_line_count = 3\nacceleration_to
;SETTING_1 pbottom = 3000\nmaterial_diameter = 1.75\nmachine_start_gcode = ; -- 
;SETTING_1 START GCODE --\n	M800        ; Custom GCODE to fire start print proce
;SETTING_1 dure\n	; -- end of START GCODE --\nacceleration_wall_x = 3000\nwirefr
;SETTING_1 ame_straight_before_down = 20\nmachine_nozzle_cool_down_speed = 2.0\n
;SETTING_1 line_width = 0.4\nlayer_height_0 = 0.2\ncool_min_layer_time_fan_speed
;SETTING_1 _max = 10\nmaterial_flow = 100\nprime_tower_line_width = 0.4\nacceler
;SETTING_1 ation_skirt = 3000\nsupport_roof_line_distance = 0.4\nsupport_join_di
;SETTING_1 stance = 2.0\njerk_infill = 20\ndraft_shield_height = 0.0\nsupport_ro
;SETTING_1 of_enable = False\nz_seam_type = shortest\ncoasting_enable = False\no
;SETTING_1 oze_shield_dist = 2\nraft_base_fan_speed = 0\ntravel_compensate_overl
;SETTING_1 apping_walls_0_enabled = True\nprime_tower_position_y = 200\ncool_min
;SETTING_1 _layer_time = 5\nbrim_line_count = 20\nacceleration_support = 3000\nl
;SETTING_1 ayer_0_z_overlap = 0.15\nsupport_enable = False\nwireframe_fall_down 
;SETTING_1 = 0.5\nraft_jerk = 20\nwireframe_printspeed_flat = 5\nacceleration_wa
;SETTING_1 ll_0 = 3000\nacceleration_print = 3000\nmagic_spiralize = False\nsupp
;SETTING_1 ort_extruder_nr = 0\nconical_overhang_enabled = False\ncooling = 0\nr
;SETTING_1 aft_surface_line_spacing = 0.3\nwall_line_width_0 = 0.4\njerk_layer_0
;SETTING_1  = 20\nswitch_extruder_prime_speed = 20\nwireframe_flow_connection = 
;SETTING_1 100\nwall_line_width = 0.4\nmeshfix_union_all_remove_holes = False\nr
;SETTING_1 aft_airgap = 0.3\nprime_tower_size = 0\nblackmagic = 0\nwireframe_pri
;SETTING_1 ntspeed_up = 5\nswitch_extruder_retraction_speeds = 20\nretraction_pr
;SETTING_1 ime_speed = 25\nmachine_show_variants = False\nsupport_z_distance = 0
;SETTING_1 .15\nraft_speed = 30.0\nbottom_layers = 6\ninfill_line_distance = 4.0
;SETTING_1 \nspeed_slowdown_layers = 2\nspeed_support_roof = 40.0\nraft_surface_
;SETTING_1 acceleration = 3000\nmachine_head_with_fans_polygon = [[-20, 10], [10
;SETTING_1 , 10], [10, -10], [-20, -10]]\ntop_thickness = 1.2\nwireframe_bottom_
;SETTING_1 delay = 0\nraft_base_thickness = 0.3\nsupport_use_towers = True\ntop_
;SETTING_1 layers = 6\nsupport_conical_angle = 30\nsupport_xy_distance = 0.7\nra
;SETTING_1 ft_interface_speed = 15.0\nsupport_roof_density = 100\nspeed_infill =
;SETTING_1  60\nmachine_extruder_count = 1\nspeed_layer_0 = 30\nraft_surface_lin
;SETTING_1 e_width = 0.3\nsupport_xy_distance_overhang = 0.2\nprime_tower_enable
;SETTING_1  = False\njerk_topbottom = 20\nsupport_conical_min_width = 5.0\ncool_
;SETTING_1 fan_full_at_height = 0.2\nacceleration_prime_tower = 3000\nsupport_ro
;SETTING_1 of_extruder_nr = 0\ncool_fan_speed_min = 100.0\nwireframe_flow = 100\
;SETTING_1 nsupport_bottom_distance = 0.1\nskirt_gap = 6\nretraction_speed = 25\
;SETTING_1 nmaterial_bed_temp_wait = True\nwireframe_printspeed_bottom = 5\nwire
;SETTING_1 frame_flow_flat = 100\nmachine_nozzle_head_distance = 3\nacceleration
;SETTING_1 _wall = 3000\ninfill_sparse_thickness = 0.2\nmachine_gcode_flavor = R
;SETTING_1 epRap\nsupport_tower_diameter = 3.0\nmachine_depth = 297\nsupport_typ
;SETTING_1 e = everywhere\nmaterial_standby_temperature = 150\nsupport_infill_ra
;SETTING_1 te = 15\nskin_line_width = 0.4\nraft_base_speed = 15.0\nsupport_infil
;SETTING_1 l_extruder_nr = 0\nmachine_head_polygon = [[-1, 1], [-1, -1], [1, -1]
;SETTING_1 , [1, 1]]\njerk_prime_tower = 20\ninfill_before_walls = True\nwirefra
;SETTING_1 me_top_delay = 0\nmagic_fuzzy_skin_enabled = False\nraft_base_acceler
;SETTING_1 ation = 3000\nretraction_amount = 6.5\nraft_base_line_width = 0.8\nwi
;SETTING_1 reframe_flat_delay = 0.1\nsupport_bottom_stair_step_height = 0.3\ntra
;SETTING_1 vel_avoid_distance = 0.625\nretraction_enable = True\nprime_tower_flo
;SETTING_1 w = 100\nraft_fan_speed = 0\nmachine_width = 210\nmagic_fuzzy_skin_po
;SETTING_1 int_density = 1.25\nraft_interface_thickness = 0.27\nsupport_xy_overr
;SETTING_1 ides_z = z_overrides_xy\nraft_interface_acceleration = 3000\nraft_bas
;SETTING_1 e_jerk = 20\ntravel_compensate_overlapping_walls_x_enabled = True\nsp
;SETTING_1 eed_topbottom = 30.0\nmagic_fuzzy_skin_thickness = 0.3\ncool_fan_spee
;SETTING_1 d_max = 100.0\nswitch_extruder_retraction_hop = 1\nswitch_extruder_re
;SETTING_1 traction_amount = 16\nraft_interface_fan_speed = 0\nplatform_adhesion
;SETTING_1  = 0\nraft_surface_layers = 2\nacceleration_enabled = False\nretracti
;SETTING_1 on_combing = all\nsupport_area_smoothing = 0.6\nretraction_extrusion_
;SETTING_1 window = 6.5\nsupport_top_distance = 0.15\ninfill = 0\nsupport_tower_
;SETTING_1 roof_angle = 65\nwireframe_roof_inset = 3\ncool_min_speed = 10\nmater
;SETTING_1 ial_extrusion_cool_down_speed = 0.5\nsupport_line_width = 0.4\nraft_i
;SETTING_1 nterface_jerk = 20\nraft_margin = 5\nmeshfix_extensive_stitching = Fa
;SETTING_1 lse\ncool_fan_enabled = True\nwireframe_strategy = compensate\njerk_s
;SETTING_1 upport_roof = 20\nsupport_minimal_diameter = 3.0\ncoasting_volume = 0
;SETTING_1 .064\ncool_fan_full_layer = 0\nresolution = 0\nraft_surface_fan_speed
;SETTING_1  = 0\nacceleration_travel = 5000\nbottom_thickness = 1.2\ncoasting_mi
;SETTING_1 n_volume = 0.8\nsupport_offset = 0.2\ninfill_overlap_mm = 0.04\ndraft
;SETTING_1 _shield_height_limitation = full\nsupport_extruder_nr_layer_0 = 0\nad
;SETTING_1 hesion_type = brim\nmaterial_print_temp_prepend = True\ngantry_height
;SETTING_1  = 99999999999\ninfill_overlap = 10\nsupport = 0\nmachine_settings = 
;SETTING_1 0\n\n
```

This behavior isn't correct isn't it?

**Comment by thopiekar** [comment]
Checked before and after deserializing the values of _name  and _id and both fit:
```
self._id:  G-code-imported-profile
self._name:  G-code-imported-profile
```

File: plugins/GCodeProfileReader/GCodeProfileReader.py [hunk 1]
```diff
@@ -72,7 +72,7 @@ def read(self, file_name):
         Logger.log("i", "Serialized the following from %s: %s" %(file_name, repr(serialized)))
 
         # Create an empty profile with the name of the G-code file
-        profile = InstanceContainer(os.path.basename(os.path.splitext(file_name)[0]))
+        profile = InstanceContainer("G-code-imported-profile")
```
Comment (fieldOfView) [inline plugins/GCodeProfileReader/GCodeProfileReader.py]: Why do you change this functionality? Now it behaves differently from importing a normal profile.
See https://github.com/Ultimaker/Cura/blob/master/cura/CuraContainerRegistry.py#L144
Also it is not translated.

**Comment by thopiekar** [comment]
Offtopic, but profile export for Ulti* devices work:
```
[general]
version = 2
name = Normal Quality
definition = ultimaker2_plus

[metadata]
type = quality
material = generic_pla_ultimaker2_plus_0.4_mm
weight = -2

[values]
speed_topbottom = 20
speed_print = 50
wall_thickness = 1.05
cool_min_speed = 10
layer_height = 0.1
top_bottom_thickness = 0.8
infill_sparse_density = 20
cool_min_layer_time = 5


```

File: plugins/GCodeProfileReader/GCodeProfileReader.py (hunk 1 above)
Comment (thopiekar) [inline plugins/GCodeProfileReader/GCodeProfileReader.py]: Oh, yes. This is wrong of course.

**Comment by thopiekar** [comment]
I rebuilt libArcus and now getting this, so still can't test SliceInfo (now) because of:
```
[35m2016-06-17 12:20:52,667[0m - [31m[47mCRITICAL[0m - [37m.../cura/CrashHandler.py (show [24]): An uncaught exception has occurred![0m
[35m2016-06-17 12:20:52,668[0m - [31m[47mCRITICAL[0m - [37m.../cura/CrashHandler.py (show [24]): TypeError: unable to convert a Python 'NoneType' object to a C++ 'bool' instance[0m
[35m2016-06-17 12:20:52,668[0m - [31m[47mCRITICAL[0m - [37m.../cura/CrashHandler.py (show [24]): An uncaught exception has occurred![0m
[35m2016-06-17 12:20:52,668[0m - [31m[47mCRITICAL[0m - [37m.../cura/CrashHandler.py (show [24]): TypeError: unable to convert a Python 'NoneType' object to a C++ 'bool' instance[0m
[35m2016-06-17 12:20:52,668[0m - [31m[47mCRITICAL[0m - [37m.../cura/CrashHandler.py (show [24]): An uncaught exception has occurred![0m
[35m2016-06-17 12:20:52,669[0m - [31m[47mCRITICAL[0m - [37m.../cura/CrashHandler.py (show [24]): TypeError: unable to convert a Python 'NoneType' object to a C++ 'bool' instance[0m
[35m2016-06-17 12:20:52,669[0m - [31m[47mCRITICAL[0m - [37m.../cura/CrashHandler.py (show [24]): An uncaught exception has occurred![0m
[35m2016-06-17 12:20:52,669[0m - [31m[47mCRITICAL[0m - [37m.../cura/CrashHandler.py (show [24]): TypeError: unable to convert a Python 'NoneType' object to a C++ 'bool' instance[0m
[35m2016-06-17 12:20:52,669[0m - [31m[47mCRITICAL[0m - [37m.../cura/CrashHandler.py (show [24]): An uncaught exception has occurred![0m
[35m2016-06-17 12:20:52,670[0m - [31m[47mCRITICAL[0m - [37m.../cura/CrashHandler.py (show [24]): TypeError: unable to convert a Python 'NoneType' object to a C++ 'bool' instance[0m
```

File: plugins/GCodeProfileReader/GCodeProfileReader.py (hunk 1 above)
Comment (thopiekar) [inline plugins/GCodeProfileReader/GCodeProfileReader.py]: is .setReadOnly(False) really needed here? The class already has got _readonly = False defined by default.

**Comment by Ghostkeeper** [comment]
The writer saves all settings, since it is unknown which profile, machine or material the user has active at the time of saving. Saving the ID of the profile wouldn't work if the profile is custom. So I figured it was best to save all settings in the profile in the g-code.

File: plugins/GCodeProfileReader/GCodeProfileReader.py (hunk 1 above)
Comment (Ghostkeeper) [inline plugins/GCodeProfileReader/GCodeProfileReader.py]: No, that shouldn't be necessary.

**Comment by thopiekar** [comment]
Ok. Why are we packaging the serialized data like this into our Gcode
files? Wouldn't it be better to save everything line by line into the
Gcode? At the end it would be easier to read..

2016-06-20 9:42 GMT+02:00 Ghostkeeper <notifications@github.com>:

**Comment by Ghostkeeper** [comment]
There is a character limit of 80 characters. Any more and Marlin firmware can't handle it! For some settings, such as start g-code, this is a problem. Instead of going through the hassle of splitting by newline and introducing additional splits if the line is too long, I opted to just escape all newlines (had to be done anyway) and cut it statically into pieces of 80 (actually a bit less since it has that prefix).

If you feel it needs improvement, you're allowed to make them but be sure to update the version number to 2 then if the format changes (so everything starts with `;SETTING_2`.

**Comment by thopiekar** [comment]
Ok, didn't know about the limit of 80 characters. Thanks!

2016-06-20 9:53 GMT+02:00 Ghostkeeper <notifications@github.com>:

**Comment by awhiemstra** [comment]
> The writer saves all settings, since it is unknown which profile, machine or material the user has active at the time of saving. Saving the ID of the profile wouldn't work if the profile is custom. So I figured it was best to save all settings in the profile in the g-code.

I see no reason why we should do that. It should be easy enough to ignore the definition at least, which would save us a lot of additional settings.

**Comment by Ghostkeeper** [comment]
As long as the user makes no additional definitions (i.e. any saved definition ID can be guaranteed to be looked up) that could work.

Remember, this feature was designed to help reproduce issues with prints.

**Comment by awhiemstra** [comment]
In that case, we can simply fail to load the profile. Anyone doing support would then need to ask for the custom definition. But at that point, you also know someone is doing something quite out of the ordinary.

**Comment by thopiekar** [comment]
So what do we want now? All settings or just the 'minimal' definitions?

**Comment by thopiekar** [comment]
In addition, what about the SliceInfo plugin? Do we still want to send a JSON will all settings? Or is it possible to send the serialized containers?

**Comment by nallath** [comment]
Serialized containers should do the trick.

File: plugins/GCodeProfileReader/GCodeProfileReader.py [hunk 2]
```diff
@@ -51,31 +55,36 @@ def read(self, file_name):
...
 
         # Perform the replacement with a regular expression.
-        serialised = pattern.sub(lambda m: GCodeProfileReader.escape_characters[re.escape(m.group(0))], serialised)
+        serialized = pattern.sub(lambda m: GCodeProfileReader.escape_characters[re.escape(m.group(0))], serialized)
+        Logger.log("i", "Serialized the following from %s: %s" %(file_name, repr(serialized)))
 
-        # Apply the changes to the current profile.
-        profile = Profile(machine_manager = Application.getInstance().getMachineManager(), read_only = False)
+        # Create an empty profile - the id will be changed later
+        profile = InstanceContainer("G-code-imported-profile")
```
Comment (fieldOfView) [inline plugins/GCodeProfileReader/GCodeProfileReader.py]: The name "G-code-imported-profile" is subsequently replaced, isn't it? Then why introduce this hardcoded name?
"" works too.

Comment (thopiekar) [inline plugins/GCodeProfileReader/GCodeProfileReader.py]: That's correct. This was added when I worked with the containers for the first time and had no idea whats happening in behind. So to be save I called the container like the exported data was named by GCodeExport.

Comment (thopiekar) [inline plugins/GCodeProfileReader/GCodeProfileReader.py]: Done

**Pull request merged**